### ISSUE TITLE
Artifact download stuck in a loop

### ISSUE THREAD

**Issue by falmanna** [eclipse/hawkbit#866]
Hello,

I have a running SWUpdate agent that is connected to local hawkbit instance. The connection is stable, and the device is able to communicate and download artifacts successfully. 

When I switched the connection to my docker based Hawkbit server running in the cloud, everything seemed to be working fine until I start an update. Once the update starts, the device seems to be running into an infinite loop. The following log shows what I mean.

`Jul 15 13:34:41 mydevice user.debug swupdate: RUN [channel_log_effective_url] : Channel's effective URL resolved to http://server/default/controller/v1/testDevice/deploymentBase/5?c=415323754
Jul 15 13:34:41 mydevice user.debug swupdate: RUN [channel_get] : Channel operation returned HTTP status code 200.
Jul 15 13:34:41 mydevice user.debug swupdate: RUN [channel_get] : Get JSON: {"id":"5","deployment":{"download":"forced","update":"forced","chunks":[{"part":"bApp","version":"1.0","name":"app","artifacts":[{"filename":"app_1.0.swu","hashes":{"sha1":"b5bfec5239d3ce42
Jul 15 13:34:41 mydevice user.debug swupdate: RUN [server_get_deployment_info] : Associated Action ID for Update Action is 5
Jul 15 13:34:41 mydevice user.info swupdate: RUN [check_action_changed] : Update classified as 'forced' by server.
Jul 15 13:34:41 mydevice user.debug swupdate: FATAL_UNKNOWN { "id" : "5", "update" : "forced"}
Jul 15 13:34:41 mydevice user.debug swupdate: RUN [server_install_update] : Iterating over JSON, key={ "part": "bApp", "version": "1.0", "name": "app", "artifacts": [ { "filename": "app_1.0.swu", "hashes": { "sha1": "b5bfec5239d3ce42e09bad9e2b947c41fde0f9a0", "md5"
Jul 15 13:34:41 mydevice user.debug swupdate: RUN [server_install_update] : Processing Update Chunk 'app', version 1.0, part bApp
Jul 15 13:34:41 mydevice user.debug swupdate: RUN [server_create_details] : Detail 0 Installing Update Chunk Artifacts.Jul 15 13:34:41 mydevice user.debug swupdate: RUN [server_create_details] : Final details: "Installing Update Chunk Artifacts."
Jul 15 13:34:41 mydevice user.debug swupdate: RUN [server_send_deployment_reply] : Reporting Installation progress for ID 5: none / proceeding / "Installing Update Chunk Artifacts."
Jul 15 13:34:41 mydevice user.debug swupdate: RUN [server_send_deployment_reply] : PUTing to http://server/default/controller/v1/testDevice/deploymentBase/5/feedback: { "id": 5, "time": "20190715T133441", "status": { "result": { "progress": { "cnt" : 0, "o
Jul 15 13:34:41 mydevice user.debug swupdate: RUN [channel_set_options] : cURL's low download speed timeout is disabled, this is most probably not what you want. Adapted it to 300s instead.
Jul 15 13:34:41 mydevice user.debug swupdate: RUN [channel_set_options] : Posted: { "id": 5, "time": "20190715T133441", "status": { "result": { "progress": { "cnt" : 0, "of" : 1 }, "finished": "none" }, "execution": "proceeding", "details" : [ "Installing Update Chunk Ar
Jul 15 13:34:41 mydevice user.debug swupdate: RUN [channel_log_effective_url] : Channel's effective URL resolved to http://server/default/controller/v1/testDevice/deploymentBase/5/feedback
Jul 15 13:34:41 mydevice user.debug swupdate: RUN [channel_post_method] : Channel put operation returned HTTP status code 200.
Jul 15 13:34:41 mydevice user.debug swupdate: RUN [server_process_update_artifact] : Iterating over JSON, key={ "filename": "app_1.0.swu", "hashes": { "sha1": "b5bfec5239d3ce42e09bad9e2b947c41fde0f9a0", "md5": "53fb944401811de98c8d99010da2bc31" }, "size": 9254400, "_lin
Jul 15 13:34:41 mydevice user.debug swupdate: RUN [server_process_update_artifact] : Processing 'app_1.0.swu' from 'https://server:8443/DEFAULT/controller/v1/testDevice/softwaremodules/2/artifacts/app_1.0.swu'
Jul 15 13:34:41 mydevice user.debug swupdate: RUN [channel_set_options] : cURL's low download speed timeout is disabled, this is most probably not what you want. Adapted it to 300s instead.
Jul 15 13:34:41 mydevice user.debug swupdate: RUN [network_thread] : Incoming network request: processing...
Jul 15 13:34:41 mydevice user.info swupdate: START Software Update started !
Jul 15 13:34:43 mydevice user.debug swupdate: UNKNOWN {"percent": 100, "msg":"Received 142B of 142B"}
Jul 15 13:34:43 mydevice user.debug swupdate: UNKNOWN {"percent": 100, "msg":"Received 142B of 142B"}
Jul 15 13:34:43 mydevice user.debug swupdate: UNKNOWN {"percent": 100, "msg":"Received 142B of 142B"}
Jul 15 13:35:04 mydevice user.debug swupdate: UNKNOWN {"percent": 0, "msg":"Received 0B of 142B"}
Jul 15 13:35:05 mydevice user.debug swupdate: UNKNOWN {"percent": 100, "msg":"Received 142B of 142B"}
Jul 15 13:35:05 mydevice user.debug swupdate: UNKNOWN {"percent": 100, "msg":"Received 142B of 142B"}
Jul 15 13:35:05 mydevice user.debug swupdate: UNKNOWN {"percent": 100, "msg":"Received 142B of 142B"}
Jul 15 13:35:29 mydevice user.debug swupdate: UNKNOWN {"percent": 0, "msg":"Received 0B of 142B"}
Jul 15 13:35:29 mydevice user.debug swupdate: UNKNOWN {"percent": 100, "msg":"Received 142B of 142B"}
Jul 15 13:35:29 mydevice user.debug swupdate: UNKNOWN {"percent": 100, "msg":"Received 142B of 142B"}
Jul 15 13:35:29 mydevice user.debug swupdate: UNKNOWN {"percent": 100, "msg":"Received 142B of 142B"}
Jul 15 13:35:48 mydevice user.debug swupdate: UNKNOWN {"percent": 0, "msg":"Received 0B of 142B"}
Jul 15 13:35:49 mydevice user.debug swupdate: UNKNOWN {"percent": 100, "msg":"Received 142B of 142B"}
Jul 15 13:35:49 mydevice user.debug swupdate: UNKNOWN {"percent": 100, "msg":"Received 142B of 142B"}
Jul 15 13:35:49 mydevice user.debug swupdate: UNKNOWN {"percent": 100, "msg":"Received 142B of 142B"}
Jul 15 13:36:07 mydevice user.debug swupdate: UNKNOWN {"percent": 0, "msg":"Received 0B of 142B"}
Jul 15 13:36:08 mydevice user.debug swupdate: UNKNOWN {"percent": 100, "msg":"Received 142B of 142B"}
Jul 15 13:36:08 mydevice user.debug swupdate: UNKNOWN {"percent": 100, "msg":"Received 142B of 142B"}
Jul 15 13:36:08 mydevice user.debug swupdate: UNKNOWN {"percent": 100, "msg":"Received 142B of 142B"}
Jul 15 13:36:26 mydevice user.debug swupdate: UNKNOWN {"percent": 0, "msg":"Received 0B of 142B"}
Jul 15 13:36:29 mydevice user.debug swupdate: UNKNOWN {"percent": 100, "msg":"Received 142B of 142B"}
Jul 15 13:36:29 mydevice user.debug swupdate: UNKNOWN {"percent": 100, "msg":"Received 142B of 142B"}
Jul 15 13:36:29 mydevice user.debug swupdate: UNKNOWN {"percent": 100, "msg":"Received 142B of 142B"}
Jul 15 13:36:50 mydevice user.debug swupdate: UNKNOWN {"percent": 0, "msg":"Received 0B of 142B"}`

The messages keep showing like this forever. I think it is not a problem with SWUpdate as it is working successfully with my local hawkbit. 
Here is the server log

`2019-07-15 13:34:40.875 DEBUG 520 --- [tp1637876401-20] o.e.h.d.rest.resource.DdiRootController : getControllerBase(testDevice)
2019-07-15 13:34:41.035 DEBUG 520 --- [tp1637876401-20] o.e.h.d.rest.resource.DdiRootController : getControllerBasedeploymentAction(testDevice,415323754)
2019-07-15 13:34:41.043 DEBUG 520 --- [tp1637876401-20] o.e.h.d.rest.resource.DdiRootController : Found an active UpdateAction for target testDevice. returning deyploment: DeploymentBase [id=5, deployment=Deployment [download=FORCED, update=FORCED, chunks=ResponseList [content=[org.eclipse.hawkbit.ddi.json.model.DdiChunk@4b4e1d15]]] actionHistory=null]
2019-07-15 13:34:41.323 DEBUG 520 --- [tp1637876401-20] o.e.h.d.rest.resource.DdiRootController : provideBasedeploymentActionFeedback for target [testDevice,5]: ActionFeedback [id=5, time=20190715T133441, status=Status [execution=PROCEEDING, result=Result [finished=NONE, progress=Progress [cnt=0, of=1]], details=[Installing Update Chunk Artifacts.]]]
2019-07-15 13:34:41.327 DEBUG 520 --- [tp1637876401-20] o.e.h.d.rest.resource.DdiRootController : Controller reported intermediate status (actionid: 5, controllerId: testDevice) as we got PROCEEDING report.
2019-07-15 13:41:20.113 WARN 520 --- [tp1637876401-13] org.atmosphere.cpr.DefaultBroadcaster : Duplicate resource 7309ad45-eeff-44b7-ba2c-a28b0428fec7. Could be caused by a dead connection not detected by your server. Replacing the old one with the fresh one
2019-07-15 13:51:25.928 WARN 520 --- [tp1637876401-85] org.atmosphere.cpr.DefaultBroadcaster : Duplicate resource 7309ad45-eeff-44b7-ba2c-a28b0428fec7. Could be caused by a dead connection not detected by your server. Replacing the old one with the fresh one`

Attached 2 images showing my server configurations.

![1](https://user-images.githubusercontent.com/3227740/61225020-68c85280-a728-11e9-9cce-c86de6f488d5.PNG)
![2](https://user-images.githubusercontent.com/3227740/61225022-69f97f80-a728-11e9-8da3-09f9c3687216.PNG)

**Comment by schabdo** [comment]
As far as I can tell from the logs above the client is only reporting intermediate feedback to the server:
```
Controller reported intermediate status (actionid: 5, controllerId: testDevice) as we got PROCEEDING report
```
Never any state change:
> It is up to the device how much intermediate feedback is provided. However, the action will be kept open until the controller on the device reports a finished (either successful or error)

**Comment by falmanna** [comment]
Does it mean that the device is not reporting progress, hence it is looping forever?

It was running perfectly on the local server, and all configs are the same except for the HTTPS. Could it be that this is somehow causing a problem?

**Comment by schabdo** [comment]
> Does it mean that the device is not reporting progress, hence it is looping forever?

Device is neither reporting "success" or "error" only feedback about update progress itself so every time device is polling it will get same update. Have a look into your client logs. There should be in both environments a call like this to finish the assigned update:

`POST /DEFAULT/controller/v1/testDevice/deploymentBase/<id>/feedback`
```json
{
    "status" : {
        "result" : {
            "finished" : "success"
        },
        "execution" : "closed"
    }
}
```

> configs are the same except for the HTTPS. Could it be that this is somehow causing a problem

I don't think so. You don't have a load-balancer nor a reverse-proxy in-between terminating the TLS for you, right?

**Comment by falmanna** [comment]
Yes, there isn't. I will do some debugging again.

**Comment by falmanna** [comment]
I printed the TRACE log, it has more info and might be of use. Please find it attached.

[swupdate-trace.log](https://github.com/eclipse/hawkbit/files/3435415/swupdate-trace.log)

**Comment by falmanna** [comment]
I finally found the issue!

Apparently, the API was returning a `download-http` URL that contains a port, and hitting that url in my server was redirecting the request to the URL without port, then downloads. 
Looks like the swupdate agent was getting redirected but it kept using the port in the hostname.

I fixed this by removing the port from the DDI api by setting:
`hawkbit.artifact.url.protocols.download-http.ref`
to
`{protocol}://{hostname}/{tenant}/controller/v1/{controllerId}/softwaremodules/{softwareModuleId}/artifacts/{artifactFileName}`
